### PR TITLE
[Merged by Bors] - Basic adaptive batching for parallel query iteration

### LINKED ISSUE #3184: Automatically specify batch size for parallel iteration
## What problem does this solve or what need does it fill?

The parallel iteration tools that currently exist in Bevy require users to manually set a 

As seen in https://github.com/bevyengine/bevy/blob/main/examples/ecs/parallel_query.rs.

## What solution would you like?

As in #3183:

1. Rename the existing methods to `par_for_each_manual` and so on.
2. Remove the ability to set the batch size for the user.
3. Set the batch size automatically using some simple and low-overhead heuristic.

## What alternative(s) have you considered?

More sophisticated strategies could be specified, but that is beyond the scope of this issue, which is focused on ergonomics.

## Additional context

Works hand-in-hand with #3183, to finish improving the default ergonomics of parallel iteration.

Smarter control over the thread count should help improve the serious performance issues noted in #2173.

### LINKED ISSUE #4798: heavy_compute bench has bad parameters
bevy 0.8-dev e5eb8376180e7571b68f99b19d77675a433e3a79

# What happened

The current heavy_compute benchmark is testing 1000 entities / 128 batch = 8 batches and using the default task pool. This makes the test overly sensitive to the number of cores on a machine.

# Desired Fix

Not sure what the best fix is here. Playing with the batch size, I was seeing the best perf at around 8 to 16. We could change the batch size or increase the number of entities or fix the number of threads to a lower number. Or vary the number of threads, Or vary the batch size.

### LINKED ISSUE #6640: `par_for_each` and `par_for_each_mut` app freeze with 100% CPU usage if an invalid batch_size is passed
## Bevy version

0.9.0

## What you did

While running a parallel query, I accidentally passed an invalid value of 0 as batch_size.

## What went wrong

This caused my application to freeze forever and apparently consume a lot of resources (I even got a blue screen of death, but I'm not sure if it was just a coincidence since it happened only once. While my application was frozen, the application was reported as using 100% CPU on the task manager.
I was expecting a panic on a clearly wrong batch_size passed as input.

## Additional information

You can easily reproduce this with this application:
```
use bevy::prelude::{App, Transform};

fn main() {
    let mut app = App::new();

    app.world.spawn(Transform::default());
    app.update();

    let mut query = app.world.query::<(&Transform,)>();

    query.par_for_each(&app.world, 0, |t| {
    })
}
```

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Fixes #3184. Fixes #6640. Fixes #4798. Using `Query::par_for_each(_mut)` currently requires a `batch_size` parameter, which affects how it chunks up large archetypes and tables into smaller chunks to run in parallel. Tuning this value is difficult, as the performance characteristics entirely depends on the state of the `World` it's being run on. Typically, users will just use a flat constant and just tune it by hand until it performs well in some benchmarks. However, this is both error prone and risks overfitting the tuning on that benchmark.

This PR proposes a naive automatic batch-size computation based on the current state of the `World`.

## Background
`Query::par_for_each(_mut)` schedules a new Task for every archetype or table that it matches. Archetypes/tables larger than the batch size are chunked into smaller tasks. Assuming every entity matched by the query has an identical workload, this makes the worst case scenario involve using a batch size equal to the size of the largest matched archetype or table. Conversely, a batch size of `max {archetype, table} size / thread count * COUNT_PER_THREAD` is likely the sweetspot where the overhead of scheduling tasks is minimized, at least not without grouping small archetypes/tables together.

There is also likely a strict minimum batch size below which the overhead of scheduling these tasks is heavier than running the entire thing single-threaded.

## Solution

- [x] Remove the `batch_size` from `Query(State)::par_for_each`  and friends.
- [x] Add a check to compute `batch_size = max {archeytpe/table} size / thread count  * COUNT_PER_THREAD`
- [x] ~~Panic if thread count is 0.~~ Defer to `for_each` if the thread count is 1 or less.
- [x] Early return if there is no matched table/archetype. 
- [x] Add override option for users have queries that strongly violate the initial assumption that all iterated entities have an equal workload.

---

## Changelog
Changed: `Query::par_for_each(_mut)` has been changed to `Query::par_iter(_mut)` and will now automatically try to produce a batch size for callers based on the current `World` state.

## Migration Guide
The `batch_size` parameter for `Query(State)::par_for_each(_mut)` has been removed. These calls will automatically compute a batch size for you. Remove these parameters from all calls to these functions.

Before:
```rust
fn parallel_system(query: Query<&MyComponent>) {
   query.par_for_each(32, |comp| {
        ...
   });
}
```

After:

```rust
fn parallel_system(query: Query<&MyComponent>) {
   query.par_iter().for_each(|comp| {
        ...
   });
}
```

### CODE REVIEW

**Comment by james7132** [comment]
To get this API surface to work, I needed to merge in #4705, so I won't take this out of draft until it or an equivalent PR gets merged.

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (hymm) [inline crates/bevy_ecs/src/query/par_iter.rs:133]: I think max / thread_count is too coarse.  Consider the case where we have one archetype, if for any reason  we don't get all the threads, the task will take twice as long to complete. This is because one of the threads will have to do 2 batches instead of just one. We probably want some type of multiplier on the thread count so we get like 4 or 5 batches per a thread. Probably needs some testing to know what that number is.

Comment (james7132) [reply crates/bevy_ecs/src/query/par_iter.rs]: Set one up at a default of 4 per thread for now. Ultimately it comes down to how much of a cost the overhead of scheduling a new task onto a thread is. If that overhead is small, we can afford to be much more granular with how we schedule each one.

As I proposed on Discord, it might be a good idea to schedule one task per thread, and just push the archetypes/batches through a SPMC channel. It'd be likely be lighter weight scheduling new tasks.

**Comment by james7132** [comment]
With #4705 merged, this is basically ready to come out of draft. I still need to update docs and do some basic benchmarking, but the base change is ready for review.

**Comment by james7132** [comment]
Now that #4489 has been merged, I was able to test this on `many_cubes` with the `check_visibility` system. This saw a drop from 2.53 ms -> 2.28 ms, a 9.09% improvement.

Below is the change in distribution on my machine, yellow is this PR, red is main.

image

**Comment by ManevilleF** [comment]
Great stuff! This will allow much more convenient parallel queries.

Small suggestion, adding the ability to override **globally** the batch size estimation with a *resource* `enum` like:

```rust
pub enum BatchSizeBehaviour {
   /// The batch size will be `artifact_count / thread_count`
   Auto,
   /// Custom batch size
   Overridden(usize),
   /// Custom 
   Custom(Box<dyn Fn(&ComputeTaskPool, &QueryFetch) -> usize>) 
}
```

That would completely remove the potential need for stuff like https://github.com/bevyengine/bevy/pull/5025

**Comment by james7132** [comment]
The estimated value comes from iterating over all matched tables or archetypes, something parallel iteration already does on the dispatch thread to begin with. If anything it makes the actual task spawning faster by getting the archetype or table metadata into the CPU cache before spawning tasks.

There are probably some degenerative cases where a query matches an enormous number of otherwise empty or near empty archetypes or tables, but I think we'd want to see some real world use cases for them first.

File: examples/ecs/parallel_query.rs
Comment (hymm) [inline examples/ecs/parallel_query.rs:52]:
```suggestion
    // The default batch size can also be overridden. 
    // In this case a batch size of 32 is chosen to limit the overhead of
    // ParallelIterator, since negating a vector is very inexpensive.
```

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (hymm) [inline crates/bevy_ecs/src/query/par_iter.rs:19]: needs some docs here

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (hymm) [inline crates/bevy_ecs/src/query/par_iter.rs:6]:
```suggestion
const BATCH_MULTIPLIER_PER_THREAD: usize = 4;
```

Comment (hymm) [reply crates/bevy_ecs/src/query/par_iter.rs:6]: probably deserves a comment explaining what changing this does.

**Review by hymm — COMMENTED** [review]
I'd like to see some perf numbers for varying the batch multiplier and number of threads to see if 4 is a good default. Ideally we'd do this against more workloads than just https://github.com/bevyengine/bevy/pull/4777.  Are there any open PR's that we could also use for benching?

**Comment by james7132** [comment]
> I'd like to see some perf numbers for varying the batch multiplier and number of threads to see if 4 is a good default. Ideally we'd do this against more workloads than just #4777. Are there any open PR's that we could also use for benching?

In retrospect, testing on that system is likely a poor choice without #4663 due to the contention on the channel. There's also #4775, which could be tested, but that's one where we might want to revisit the parallelism strategy too due to the way the system runs. Unfortunately there aren't many options for an idealized case here right now, at least not until #4663 is merged.

**Comment by james7132** [comment]
With #4663 merged, the timing on each batch is more consistent now and is more reflective of the "equal amount of work per entity" than before. Here are some updated timings while varying the batches per thread:

|attempt|timing|
|:--|:--|
|main (static 1024)|1.67ms|
|1 batch per thread|996.43us|
|2 batches per thread|1.05us|
|4 batches per thread|1.26ms|
|8 batches per thread|2.07ms|

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (james7132) [reply crates/bevy_ecs/src/query/par_iter.rs]: No longer relevant as this is no longer a constant, this is configurable, and defaults to 1. Should be better documented too.

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (james7132) [reply crates/bevy_ecs/src/query/par_iter.rs]: Done.

**Review by ManevilleF — APPROVED** [review]

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/par_iter.rs:180]: Can we use the new `ExactSizeIterator` APIs here? Do we want to?

**Comment by superdump** [comment]
@james7132 @hymm what CPUs do you both have?

**Comment by james7132** [comment]
5950x. By default, ComputeTaskPool gets assigned 24 threads.

File: examples/ecs/parallel_query.rs
Comment (superdump) [inline examples/ecs/parallel_query.rs:57]: I like that we can configure different strategies to do overrides like this.

**Review by superdump — APPROVED** [review]
I think this looks good as far as the batching is concerned. Simple, clear, and I like that the batching strategy can be overridden.

NOTE: I have not looked in detail at the parallel iterator implementations as all that kind of thing is still a scary black box of concerns to me. So, I think outstanding things to do are have someone who is strong on ECS / safety stuff review the parallel iterator changes, and maybe some more benchmarking.

**Comment by hymm** [comment]
5600x

The only reason I haven't approved this yet is that james7132 have different values for best batches/thread. We need some more data points from other cpus to determine a good default.

**Comment by superdump** [comment]
Results from an M1 Max, running `many_cubes -- sphere` for 1500 frames:
Screenshot 2022-07-18 at 10 53 18

With this patch:
```patch
diff --git a/crates/bevy_ecs/src/query/state.rs b/crates/bevy_ecs/src/query/state.rs
index 60643ace4..014ad7ab5 100644
--- a/crates/bevy_ecs/src/query/state.rs
+++ b/crates/bevy_ecs/src/query/state.rs
@@ -936,15 +936,15 @@ impl<Q: WorldQuery, F: WorldQuery> QueryState<Q, F> {
                                 func(item);
                             }
                         };
-                        #[cfg(feature = "trace")]
-                        let span = bevy_utils::tracing::info_span!(
-                            "par_for_each",
-                            query = std::any::type_name::<Q>(),
-                            filter = std::any::type_name::<F>(),
-                            count = len,
-                        );
-                        #[cfg(feature = "trace")]
-                        let task = task.instrument(span);
+                        // #[cfg(feature = "trace")]
+                        // let span = bevy_utils::tracing::info_span!(
+                        //     "par_for_each",
+                        //     query = std::any::type_name::<Q>(),
+                        //     filter = std::any::type_name::<F>(),
+                        //     count = len,
+                        // );
+                        // #[cfg(feature = "trace")]
+                        // let task = task.instrument(span);
                         scope.spawn(task);
                         offset += batch_size;
                     }
@@ -979,15 +979,15 @@ impl<Q: WorldQuery, F: WorldQuery> QueryState<Q, F> {
                             }
                         };
 
-                        #[cfg(feature = "trace")]
-                        let span = bevy_utils::tracing::info_span!(
-                            "par_for_each",
-                            query = std::any::type_name::<Q>(),
-                            filter = std::any::type_name::<F>(),
-                            count = len,
-                        );
-                        #[cfg(feature = "trace")]
-                        let task = task.instrument(span);
+                        // #[cfg(feature = "trace")]
+                        // let span = bevy_utils::tracing::info_span!(
+                        //     "par_for_each",
+                        //     query = std::any::type_name::<Q>(),
+                        //     filter = std::any::type_name::<F>(),
+                        //     count = len,
+                        // );
+                        // #[cfg(feature = "trace")]
+                        // let task = task.instrument(span);
 
                         scope.spawn(task);
                         offset += batch_size;
diff --git a/crates/bevy_render/src/view/visibility/mod.rs b/crates/bevy_render/src/view/visibility/mod.rs
index c1bffec3f..e0f0eede3 100644
--- a/crates/bevy_render/src/view/visibility/mod.rs
+++ b/crates/bevy_render/src/view/visibility/mod.rs
@@ -5,7 +5,7 @@ pub use render_layers::*;
 
 use bevy_app::{CoreStage, Plugin};
 use bevy_asset::{Assets, Handle};
-use bevy_ecs::prelude::*;
+use bevy_ecs::{prelude::*, query::BatchingStrategy};
 use bevy_reflect::std_traits::ReflectDefault;
 use bevy_reflect::Reflect;
 use bevy_transform::components::GlobalTransform;
@@ -174,51 +174,63 @@ pub fn check_visibility(
         let view_mask = maybe_view_mask.copied().unwrap_or_default();
 
         visible_entities.entities.clear();
-        visible_entity_query.p1().par_iter().for_each_mut(
-            |(
-                entity,
-                visibility,
-                mut computed_visibility,
-                maybe_entity_mask,
-                maybe_aabb,
-                maybe_no_frustum_culling,
-                maybe_transform,
-            )| {
-                if !visibility.is_visible {
-                    return;
-                }
-
-                let entity_mask = maybe_entity_mask.copied().unwrap_or_default();
-                if !view_mask.intersects(&entity_mask) {
-                    return;
-                }
-
-                // If we have an aabb and transform, do frustum culling
-                if let (Some(model_aabb), None, Some(transform)) =
-                    (maybe_aabb, maybe_no_frustum_culling, maybe_transform)
-                {
-                    let model = transform.compute_matrix();
-                    let model_sphere = Sphere {
-                        center: model.transform_point3a(model_aabb.center),
-                        radius: (Vec3A::from(transform.scale) * model_aabb.half_extents).length(),
-                    };
-                    // Do quick sphere-based frustum culling
-                    if !frustum.intersects_sphere(&model_sphere, false) {
+        visible_entity_query
+            .p1()
+            .par_iter()
+            .batching_strategy(
+                BatchingStrategy::new().batches_per_thread(
+                    std::env::args()
+                        .nth(3)
+                        .and_then(|arg| arg.parse::<usize>().ok())
+                        .unwrap(),
+                ),
+            )
+            .for_each_mut(
+                |(
+                    entity,
+                    visibility,
+                    mut computed_visibility,
+                    maybe_entity_mask,
+                    maybe_aabb,
+                    maybe_no_frustum_culling,
+                    maybe_transform,
+                )| {
+                    if !visibility.is_visible {
                         return;
                     }
-                    // If we have an aabb, do aabb-based frustum culling
-                    if !frustum.intersects_obb(model_aabb, &model, false) {
+
+                    let entity_mask = maybe_entity_mask.copied().unwrap_or_default();
+                    if !view_mask.intersects(&entity_mask) {
                         return;
                     }
-                }
-
-                computed_visibility.is_visible = true;
-                let cell = thread_queues.get_or_default();
-                let mut queue = cell.take();
-                queue.push(entity);
-                cell.set(queue);
-            },
-        );
+
+                    // If we have an aabb and transform, do frustum culling
+                    if let (Some(model_aabb), None, Some(transform)) =
+                        (maybe_aabb, maybe_no_frustum_culling, maybe_transform)
+                    {
+                        let model = transform.compute_matrix();
+                        let model_sphere = Sphere {
+                            center: model.transform_point3a(model_aabb.center),
+                            radius: (Vec3A::from(transform.scale) * model_aabb.half_extents)
+                                .length(),
+                        };
+                        // Do quick sphere-based frustum culling
+                        if !frustum.intersects_sphere(&model_sphere, false) {
+                            return;
+                        }
+                        // If we have an aabb, do aabb-based frustum culling
+                        if !frustum.intersects_obb(model_aabb, &model, false) {
+                            return;
+                        }
+                    }
+
+                    computed_visibility.is_visible = true;
+                    let cell = thread_queues.get_or_default();
+                    let mut queue = cell.take();
+                    queue.push(entity);
+                    cell.set(queue);
+                },
+            );
 
         for cell in thread_queues.iter_mut() {
             visible_entities.entities.append(cell.get_mut());
diff --git a/examples/stress_tests/many_cubes.rs b/examples/stress_tests/many_cubes.rs
index 6ced37ee1..844584f2d 100644
--- a/examples/stress_tests/many_cubes.rs
+++ b/examples/stress_tests/many_cubes.rs
@@ -11,13 +11,27 @@
 //! `cargo run --example many_cubes --release sphere`
 
 use bevy::{
+    core::TaskPoolThreadAssignmentPolicy,
     diagnostic::{FrameTimeDiagnosticsPlugin, LogDiagnosticsPlugin},
     math::{DVec2, DVec3},
     prelude::*,
 };
 
 fn main() {
+    let n_compute_threads = std::env::args()
+        .nth(2)
+        .and_then(|arg| arg.parse::<usize>().ok())
+        .unwrap();
     App::new()
+        .insert_resource(DefaultTaskPoolOptions {
+            compute: TaskPoolThreadAssignmentPolicy {
+                min_threads: n_compute_threads,
+                max_threads: n_compute_threads,
+                percent: 1.0,
+            },
+            ..default()
+        })
         .add_plugins(DefaultPlugins)
         .add_plugin(FrameTimeDiagnosticsPlugin::default())
         .add_plugin(LogDiagnosticsPlugin::default())
```

**Comment by inodentry** [comment]
From an earlier comment:

> As I proposed on Discord, it might be a good idea to schedule one task per thread, and just push the archetypes/batches through a SPMC channel. It'd be likely be lighter weight than scheduling a new task per archetype or batch.

Why wasn't this implemented instead? Why does this PR keep the old approach?

**Comment by hymm** [comment]
> As I proposed on Discord, it might be a good idea to schedule one task per thread, and just push the archetypes/batches through a SPMC channel. It'd be likely be lighter weight than scheduling a new task per archetype or batch.

It probably doesn't work well if all the threads aren't available, since there's no task stealing in this model.

**Comment by james7132** [comment]
> Why wasn't this implemented instead? Why does this PR keep the old approach?

Actually completely slipped my mind why I didn't. Even then, I would like to avoid conflating the performance differences of both adaptive batching and a channel based approach in a single PR.

> It probably doesn't work well if all the threads aren't available, since there's no task stealing in this model.

The work stealing here definitely helps smooth out the workload across each thread. However, even with that said, the idea of decreasing contention on the global executor, using a fixed sized channel to minimize allocator time, fewer atomics per batch, and potentially not needing to yield time to park threads is pretty appealing.

**Review by hymm — APPROVED** [review]
Code looks good to me. Not sure about the batches per thread default of 1, but it seems to be fine given the data that we have. It can be fine tuned in other pr's after we get more usages of par_for_each in the engine.

**Review by maniwani — APPROVED** [review]

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/par_iter.rs:25]: Doc strings for each of these parameters would be nice; especially if you can suggest how and why you might vary these.

**Review by alice-i-cecile — APPROVED** [review]
I really like the UX and the config struct pattern. More docs are always nice but I won't block on it.

File: crates/bevy_ecs/src/query/par_iter.rs
Comment (james7132) [reply crates/bevy_ecs/src/query/par_iter.rs:31]: Done.

**Comment by james7132** [comment]
I was holding off from merging this, even with multiple ECS SME approval, as there was a regression with `check_visibility` in `many_foxes` due to it generating too many small tasks, (216us -> 316us), though the other internally parallel systems showed no significant changes.

Checking against `many_cubes`, this does have a much higher benefit, showing a near 3x speedup in `PostUpdate` (2.05ms -> 896.43us).

image

This does seem to suggest we're oversplitting on worlds with too many archetypes or too many threads with small archetypes, while also delivering on reducing scheduling overhead on times when we're underclustering the batches before.

As mentioned before, in the interest of both improving UX, probably best not to block this PR any further and make changes to the heuristics and tuning in followup PRs.

bors r+

**Pull request closed without merging**
